### PR TITLE
Tests for token service

### PR DESCRIPTION
**Pull request by dimaatkaev**
### Please read and mark the following check list before creating a pull request:

 - [] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:

### CODE REVIEW

File: src/framework/auth/services/token.service.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ import { deepExtend, getDeepFromObject, urlBase64Decode } from '../helpers';
 @Injectable()
 export class NbAuthSimpleToken {
 
+  // TODO Why empty string. localStorage will set a null value anyway
   protected token: string = '';
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: let's set it to null then by default

File: src/framework/auth/services/token.service.ts [hunk 2]
```diff
@@ -28,6 +29,14 @@ export class NbAuthSimpleToken {
   getValue() {
     return this.token;
   }
+
+  /**
+   * Is current token could be processed
+   * @returns {boolean}
+   */
+  isValid(): boolean {
+    return this.token == null;
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: `return !!this.token`

File: src/framework/auth/services/token.service.ts [hunk 3]
```diff
@@ -71,6 +80,18 @@ export class NbAuthJWTToken extends NbAuthSimpleToken {
 
     return date;
   }
+
+  /**
+   * Is data expired
+   * @returns {boolean}
+   */
+  isValid(): boolean {
+    const expDate: Date = this.getTokenExpDate();
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: could be much simpler as `return super.isValid() && new Date() < this.getTokenExpDate();`

File: src/framework/auth/services/token.service.ts [hunk 4]
```diff
@@ -105,6 +126,7 @@ export class NbTokenService {
         return observableOf(this.tokenWrapper);
       },
 
+      // FIXME is it possible to unify interface
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: let's use TODO as a unified todo's tag

File: src/framework/auth/services/token.spec.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      tokenService.set(testTokenValue)
+
+      expect(localStorage.getItem(testTokenKey)).toEqual(testTokenValue);
+    },
+  );
+
+  it('set null as string', () => {
+      tokenService.set(null);
+
+      expect(localStorage.getItem(testTokenKey)).toEqual('null');
```
Comment (nnixaa) [inline src/framework/auth/services/token.spec.ts]: .toEqual('null') looks strange

File: src/framework/auth/services/token.spec.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    tokenService.get().subscribe(resp => {
+      expect(resp['token']).toEqual(testTokenValue);
+    });
+  });
+
+  it('clear remove token', () => {
+    localStorage.setItem(testTokenKey, testTokenValue);
+
+    tokenService.clear();
```
Comment (nnixaa) [inline src/framework/auth/services/token.spec.ts]: shouldn't we subscribe?

File: src/framework/auth/services/token.spec.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    expect(localStorage.getItem(testTokenKey)).toBeNull();
+  });
+
+  it('clear remove token only', () => {
+    localStorage.setItem(testTokenKey, testTokenValue);
+    localStorage.setItem(testTokenKey + '2', testTokenValue);
+
+    tokenService.clear();
```
Comment (nnixaa) [inline src/framework/auth/services/token.spec.ts]: shouldn't we subscribe?

File: src/framework/auth/services/token.spec.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    tokenService.get().subscribe(resp => {
+      expect(resp['token']).toBeNull();
+    });
+  });
+
+  it('should return correct value', () => {
+    localStorage.setItem(testTokenKey, testTokenValue);
+
+    tokenService.get().subscribe(resp => {
+      expect(resp['token']).toEqual(testTokenValue);
```
Comment (dimaatkaev) [inline src/framework/auth/services/token.spec.ts]: Probably value from token should be taken in this way: `token.getValue()`

File: src/framework/auth/services/token.service.ts [hunk 9]
```diff
@@ -105,9 +129,17 @@ export class NbTokenService {
         return observableOf(this.tokenWrapper);
       },
 
+      // TODO is it possible to unify interface
       setter: (token: string | NbAuthSimpleToken): Observable<null> => {
-        const raw = token instanceof NbAuthSimpleToken ? token.getValue() : token;
-        localStorage.setItem(this.getConfigValue('token.key'), raw);
+        let tokenValue;
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: Let's get gid of one `else` clause for simplicity 
```
if (!(token instanceof NbAuthSimpleToken)) {
  this.tokenWrapper.setValue(token); // ideally in future, we do this `token = this.tokenWrapper.createToken(token);`
  token = this.tokenWrapper;
}
localStorage.setItem(this.getConfigValue('token.key'), token.toString());
```

File: src/framework/auth/services/token.service.ts [hunk 10]
```diff
@@ -15,7 +15,7 @@ import { deepExtend, getDeepFromObject, urlBase64Decode } from '../helpers';
 @Injectable()
 export class NbAuthSimpleToken {
```
Comment (nnixaa) [inline src/framework/auth/services/token.service.ts]: We also would need tests for the token classes also.

File: src/framework/auth/services/token.service.ts (hunk 1 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: fixed

File: src/framework/auth/services/token.service.ts (hunk 2 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: fixed

File: src/framework/auth/services/token.service.ts (hunk 3 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: fixed

File: src/framework/auth/services/token.service.ts (hunk 4 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: fixed

File: src/framework/auth/services/token.spec.ts (hunk 5 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.spec.ts]: fixed

File: src/framework/auth/services/token.spec.ts (hunk 6 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.spec.ts]: fixed

File: src/framework/auth/services/token.spec.ts (hunk 7 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.spec.ts]: fixed

File: src/framework/auth/services/token.spec.ts (hunk 8 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.spec.ts]: fixed

File: src/framework/auth/services/token.service.ts (hunk 9 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: fixed

File: src/framework/auth/services/token.service.ts (hunk 10 above)
Comment (dimaatkaev) [reply src/framework/auth/services/token.service.ts]: done

**Comment by nnixaa** [comment]
closing as refactored in https://github.com/akveo/nebular/pull/235

**Pull request closed without merging**